[BUG] Workspace `file:` dependency on a non-workspace package is silently skipped with `install-strategy=linked`

## Is there an existing issue for this?

- [x] I have searched the existing issues

## This issue exists in the latest npm version

- [x] I am using the latest npm

## Current Behavior

When a **workspace** declares a `file:` dependency that points to a **non-workspace** local package (a directory outside the project's `workspaces` globs) and the project is installed with `install-strategy=linked`, the dependency is **silently skipped**. `npm install` reports `up to date` / `added 0 packages`, no `node_modules` entry is created for the dep, and the workspace cannot resolve it at runtime (`require` throws `MODULE_NOT_FOUND`). No error or warning is printed.

The key trigger is that the `file:` **target is not itself a workspace**. A workspace `file:` dep that points to another workspace (e.g. `"@scope/b": "file:../b"` where `b` is under `packages/*` — the common monorepo case, e.g. Gutenberg) works fine, because the target is matched by `n.isWorkspace`. The bug only surfaces for `file:` targets outside the workspace set — which is exactly the [#9115](https://github.com/npm/cli/issues/9115) case (`npm --workspace=<ws> link <external-path>`), and any workspace depending on a local non-workspace folder.

This is also distinct from the already-fixed [#9166](https://github.com/npm/cli/issues/9166) / [#9210](https://github.com/npm/cli/issues/9210) (which crashed with `ENOENT` for `file:` deps in a **non-workspace** project). Those now work; this is the residual case where the consumer is a workspace and the target is not.

So `install-strategy=linked` is not a viable workaround for #9115.

### What works vs what fails (linked strategy)

| Scenario | Result |
| --- | --- |
| Workspace `file:` dep → **non-workspace** local package | ❌ silently skipped (not installed, no error) |
| Workspace `file:` dep → **another workspace** (e.g. Gutenberg `file:../i18n`) | ✅ installed + resolves |
| **Root**-declared `file:` dep (workspaces present, any target) | ✅ installed + resolves |
| `file:` dep declared by root in a **non-workspace** project | ✅ installed + resolves |
| **Registry** dep declared by a workspace | ✅ installed into `node_modules/.store` + symlinked into the workspace |

## Expected Behavior

A `file:` dependency declared by a workspace should be symlinked into that workspace's `node_modules` under `install-strategy=linked`, identical to a registry dependency and identical to the hoisted strategy. If it genuinely cannot be installed, npm must error — not silently report success.

## Steps To Reproduce

```bash
rm -rf /tmp/repro9115 && mkdir -p /tmp/repro9115/packages/ws-a /tmp/repro9115/local-dep

cat > /tmp/repro9115/package.json << 'EOF'
{ "name": "mono", "version": "1.0.0", "workspaces": ["packages/*"] }
EOF

cat > /tmp/repro9115/local-dep/package.json << 'EOF'
{ "name": "local-dep", "version": "1.0.0" }
EOF

cat > /tmp/repro9115/packages/ws-a/package.json << 'EOF'
{ "name": "ws-a", "version": "1.0.0", "dependencies": { "local-dep": "file:../../local-dep" } }
EOF

cd /tmp/repro9115
npm install --install-strategy=linked
# => "up to date, audited 5 packages" — but nothing is installed

find . -type d -name node_modules          # => (no node_modules at all)
node -e "require.resolve('local-dep/package.json', { paths: ['/tmp/repro9115/packages/ws-a'] })"
# => Error: Cannot find module 'local-dep/package.json'
```

### Works with hoisted strategy (control)

```bash
rm -rf /tmp/repro9115/node_modules /tmp/repro9115/package-lock.json
cd /tmp/repro9115
npm install
node -e "console.log(require.resolve('local-dep/package.json', { paths: ['/tmp/repro9115/packages/ws-a'] }))"
# => /private/tmp/repro9115/local-dep/package.json
```

## Environment

- npm: 12.0.0-pre.1 (reproduced on `latest`; line unchanged across recent branches)
- Node.js: v24.17.0
- OS Name: macOS (Darwin 25.5.0)
- npm config:
```ini
install-strategy=linked
```

Related: #9115, #9166, #9210.

--- a/workspaces/arborist/lib/arborist/isolated-reifier.js
+++ b/workspaces/arborist/lib/arborist/isolated-reifier.js
@@ -228,8 +228,9 @@ module.exports = cls => class IsolatedReifier extends cls {
     }
 
     // local `file:` deps (non-workspace fsChildren) should be treated as local dependencies, not external, so they get symlinked directly instead of being extracted into the store.
-    // fsChildren only exists on the root ideal tree, so also consult it directly to catch file: targets depended on by a workspace (node.fsChildren is undefined there).
-    const isLocal = (n) => n.isWorkspace || node.fsChildren?.has(n) || this.idealTree.fsChildren?.has(n)
+    // A file: dep surfaces as a Link edge whose resolved spec starts with file:; detect it from the edge so the target is treated as local even when it is absent from idealTree.fsChildren (a workspace consumer, or a target outside the repo root via npm link).
+    const fileLinkTargets = new Set(edges.filter(e => e.to?.isLink && e.to.resolved?.startsWith('file:')).map(e => e.to.target))
+    const isLocal = (n) => n.isWorkspace || node.fsChildren?.has(n) || fileLinkTargets.has(n)
     const optionalDeps = edges.filter(edge => edge.optional).map(edge => edge.to.target)
 
     // Optional peers declared only in peerDependenciesMeta (e.g. `@types/react`) have no edge, so the materialization above misses them.

--- a/workspaces/arborist/lib/arborist/isolated-reifier.js
+++ b/workspaces/arborist/lib/arborist/isolated-reifier.js
@@ -228,7 +228,8 @@ module.exports = cls => class IsolatedReifier extends cls {
     }
 
     // local `file:` deps (non-workspace fsChildren) should be treated as local dependencies, not external, so they get symlinked directly instead of being extracted into the store.
-    const isLocal = (n) => n.isWorkspace || node.fsChildren?.has(n)
+    // fsChildren only exists on the root ideal tree, so also consult it directly to catch file: targets depended on by a workspace (node.fsChildren is undefined there).
+    const isLocal = (n) => n.isWorkspace || node.fsChildren?.has(n) || this.idealTree.fsChildren?.has(n)
     const optionalDeps = edges.filter(edge => edge.optional).map(edge => edge.to.target)
 
     // Optional peers declared only in peerDependenciesMeta (e.g. `@types/react`) have no edge, so the materialization above misses them.

--- a/workspaces/arborist/test/isolated-mode.js
+++ b/workspaces/arborist/test/isolated-mode.js
@@ -1704,6 +1704,54 @@ tap.test('workspace file: dependency on a non-workspace local package with linke
   t.ok(setupRequire(path.join(dir, 'packages', 'ws-a'))('local-dep'), 'local-dep can be required from the workspace')
 })
 
+tap.test('workspace file: dependency on a package outside the repo root with linked strategy', async t => {
+  // Regression test for the out-of-repo variant of https://github.com/npm/cli/issues/9589 (the real `npm --workspace link <external>` case, https://github.com/npm/cli/issues/9115).
+  // A workspace file: dep whose target resolves OUTSIDE the repo root was silently skipped.
+  // The target is not in idealTree.fsChildren, so the fix must detect it from the file: link edge.
+  const graph = {
+    registry: [],
+    root: {
+      name: 'mono',
+      version: '1.0.0',
+    },
+    workspaces: [
+      { name: 'ws-a', version: '1.0.0', dependencies: { 'ext-pkg': 'file:../../../ext-pkg' } },
+    ],
+  }
+
+  const { dir, registry } = await getRepo(graph)
+
+  // Create the external package OUTSIDE the repo root
+  const extDir = path.join(path.dirname(dir), 'ext-pkg')
+  fs.mkdirSync(extDir, { recursive: true })
+  fs.writeFileSync(path.join(extDir, 'package.json'), JSON.stringify({
+    name: 'ext-pkg',
+    version: '1.0.0',
+  }))
+  fs.writeFileSync(path.join(extDir, 'index.js'), "module.exports = 'ext-pkg'")
+
+  const cache = fs.mkdtempSync(`${getTempDir()}/test-`)
+  const arborist = new Arborist({ path: dir, registry, packumentCache: new Map(), cache })
+  await arborist.reify({ installStrategy: 'linked' })
+
+  // The file dep should be symlinked into the workspace's node_modules
+  const linkPath = path.join(dir, 'packages', 'ws-a', 'node_modules', 'ext-pkg')
+  const stat = fs.lstatSync(linkPath)
+  t.ok(stat.isSymbolicLink(), 'ext-pkg is a symlink in the workspace node_modules')
+
+  // The symlink should resolve to the actual external directory
+  t.equal(fs.realpathSync(linkPath), fs.realpathSync(extDir), 'symlink points to the correct external directory')
+
+  // It must be symlinked directly, not extracted into the store
+  const storePath = path.join(dir, 'node_modules', '.store')
+  if (fs.existsSync(storePath)) {
+    t.notOk(fs.readdirSync(storePath).some(e => e.startsWith('ext-pkg@')), 'ext-pkg is NOT in the store')
+  }
+
+  // The package should be requireable from inside the workspace
+  t.ok(setupRequire(path.join(dir, 'packages', 'ws-a'))('ext-pkg'), 'ext-pkg can be required from the workspace')
+})
+
 tap.test('subsequent linked install is a no-op', async t => {
   const graph = {
     registry: [

--- a/workspaces/arborist/test/isolated-mode.js
+++ b/workspaces/arborist/test/isolated-mode.js
@@ -1657,6 +1657,53 @@ tap.test('npm link (external file: dep) with linked strategy', async t => {
   t.notOk(storeEntries.some(e => e.startsWith('external-pkg@')), 'external-pkg is NOT in the store')
 })
 
+tap.test('workspace file: dependency on a non-workspace local package with linked strategy', async t => {
+  // Regression test for https://github.com/npm/cli/issues/9589
+  // A workspace declaring a file: dep on a local package that is NOT itself a workspace was silently skipped: no symlink, no error.
+  const graph = {
+    registry: [],
+    root: {
+      name: 'mono',
+      version: '1.0.0',
+    },
+    workspaces: [
+      { name: 'ws-a', version: '1.0.0', dependencies: { 'local-dep': 'file:../../local-dep' } },
+    ],
+  }
+
+  const { dir, registry } = await getRepo(graph)
+
+  // Create the non-workspace local package on disk, outside the workspaces globs
+  const depDir = path.join(dir, 'local-dep')
+  fs.mkdirSync(depDir, { recursive: true })
+  fs.writeFileSync(path.join(depDir, 'package.json'), JSON.stringify({
+    name: 'local-dep',
+    version: '1.0.0',
+  }))
+  fs.writeFileSync(path.join(depDir, 'index.js'), "module.exports = 'local-dep'")
+
+  const cache = fs.mkdtempSync(`${getTempDir()}/test-`)
+  const arborist = new Arborist({ path: dir, registry, packumentCache: new Map(), cache })
+  await arborist.reify({ installStrategy: 'linked' })
+
+  // The file dep should be symlinked into the workspace's node_modules
+  const linkPath = path.join(dir, 'packages', 'ws-a', 'node_modules', 'local-dep')
+  const stat = fs.lstatSync(linkPath)
+  t.ok(stat.isSymbolicLink(), 'local-dep is a symlink in the workspace node_modules')
+
+  // The symlink should resolve to the actual local directory
+  t.equal(fs.realpathSync(linkPath), fs.realpathSync(depDir), 'symlink points to the correct local directory')
+
+  // It must be symlinked directly, not extracted into the store
+  const storePath = path.join(dir, 'node_modules', '.store')
+  if (fs.existsSync(storePath)) {
+    t.notOk(fs.readdirSync(storePath).some(e => e.startsWith('local-dep@')), 'local-dep is NOT in the store')
+  }
+
+  // The package should be requireable from inside the workspace
+  t.ok(setupRequire(path.join(dir, 'packages', 'ws-a'))('local-dep'), 'local-dep can be required from the workspace')
+})
+
 tap.test('subsequent linked install is a no-op', async t => {
   const graph = {
     registry: [

